github-status/handler_test.go

```diff
@@ -0,0 +1,304 @@
+package function
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/openfaas/openfaas-cloud/sdk"
+)
+
+func TestBuildURLWithoutPrettyURL_WithSlash(t *testing.T) {
+	os.Setenv("gateway_public_url", "http://localhost:8080")
+	os.Setenv("gateway_pretty_url", "")
+
+	event := &sdk.Event{
+		Owner:   "alexellis",
+		Service: "tester",
+	}
+
+	val := buildPublicStatusURL("success", sdk.FunctionContext(event.Service), event)
+	want := "http://localhost:8080/function/alexellis-tester"
+
+	if val != want {
+		t.Errorf("building PublicURL: want %s, got %s", want, val)
+		t.Fail()
+	}
+}
+
+func TestBuildURLWithoutPrettyURL_WithOutSlash(t *testing.T) {
+	os.Setenv("gateway_public_url", "http://localhost:8080")
+	os.Setenv("gateway_pretty_url", "")
+
+	event := &sdk.Event{
+		Owner:   "alexellis",
+		Service: "tester",
+	}
+
+	val := buildPublicStatusURL("success", sdk.FunctionContext(event.Service), event)
+	want := "http://localhost:8080/function/alexellis-tester"
+
+	if val != want {
+		t.Errorf("building PublicURL: want %s, got %s", want, val)
+		t.Fail()
+	}
+}
+
+func TestBuildURLWithPrettyURL(t *testing.T) {
+	os.Setenv("gateway_public_url", "http://localhost:8080")
+	os.Setenv("gateway_pretty_url", "https://user.openfaas-cloud.com/function")
+
+	event := &sdk.Event{
+		Owner:   "alexellis",
+		Service: "tester",
+	}
+
+	val := buildPublicStatusURL("success", sdk.FunctionContext(event.Service), event)
+	want := "https://alexellis.openfaas-cloud.com/tester"
+
+	if val != want {
+		t.Errorf("building PublicURL: want %s, got %s", want, val)
+		t.Fail()
+	}
+}
+
+func TestBuildURLWithUndefinedStatusGivesOriginalURL(t *testing.T) {
+
+	event := &sdk.Event{
+		Owner:   "alexellis",
+		Service: "tester",
+		URL:     "http://original-value.local",
+	}
+
+	val := buildPublicStatusURL("not-supported", sdk.FunctionContext(event.Service), event)
+	want := event.URL
+
+	if val != want {
+		t.Errorf("building PublicURL: want %s, got %s", want, val)
+		t.Fail()
+	}
+}
+
+func TestTokenValidator(t *testing.T) {
+	testTokens := map[string]bool{
+		"token with space":                                false,
+		"token$With@special=Char":                         false,
+		"v1.afbce39asdasd8be30123317cef123321ae991cf40f7": true,
+		"token=v1.afbce39asdasd8be30123":                  false,
+		" ": false}
+	for token, result := range testTokens {
+		if sdk.ValidToken(token) != result {
+			t.Errorf("validating token %s: want %v, got %v", token, result, !result)
+		}
+	}
+}
+
+func TestTokenMarshal(t *testing.T) {
+	testTokens := map[string]string{
+		"token with space":                                sdk.EmptyAuthToken,
+		"token$With@special=Char":                         sdk.EmptyAuthToken,
+		"v1.afbce39asdasd8be30123317cef123321ae991cf40f7": "v1.afbce39asdasd8be30123317cef123321ae991cf40f7",
+		"token=v1.afbce39asdasd8be30123":                  sdk.EmptyAuthToken,
+		" ": sdk.EmptyAuthToken}
+	for token, result := range testTokens {
```

Comment:
Use sub-tests here.

Reply:
I have added a separate Test case to check the `UnmarshalStatus` Error cases, any other scenario are getting covered here.

Reply:
You need to use a sub-test, it's something like calling `t.Test()` and passing in a scenario name. Please refer to test tables too. @rgee0 can point you in the right direction.

Reply:
Pop over to [Derek](https://github.com/alexellis/derek/blob/master/commentHandler_test.go) and have a look at the tests in [commentHandler](https://github.com/alexellis/derek/blob/master/commentHandler_test.go).

Basically, put all your test cases into a struct then loop through that to test each scenario.  Using a test.title helps in tracking down any fails.

Reply:
Updated in the new commit
https://github.com/openfaas/openfaas-cloud/pull/83/files#diff-b3c2497a977c4257af245d1f1f78474bR82